libraries/AP_GPS/AP_GPS_SBP2.cpp

```diff
@@ -0,0 +1,494 @@
+                    parser_state.n_read = 0;
+                    parser_state.state = sbp_parser_state_t::GET_CRC;
+                }
+                break;
+
+            case sbp_parser_state_t::GET_CRC:
+                *((uint8_t*)&(parser_state.crc) + parser_state.n_read) = temp;
+                parser_state.n_read += 1;
+                if (parser_state.n_read >= 2) {
+                    parser_state.state = sbp_parser_state_t::WAITING;
+
+                    crc = crc16_ccitt((uint8_t*)&(parser_state.msg_type), 2, 0);
+                    crc = crc16_ccitt((uint8_t*)&(parser_state.sender_id), 2, crc);
+                    crc = crc16_ccitt(&(parser_state.msg_len), 1, crc);
+                    crc = crc16_ccitt(parser_state.msg_buff, parser_state.msg_len, crc);
+                    if (parser_state.crc == crc) {
+                        _sbp_process_message();
+                    } else {
+                        Debug("CRC Error Occurred!");
+                        crc_error_counter += 1;
+                    }
+
+                    parser_state.state = sbp_parser_state_t::WAITING;
+                }
+                break;
+
+            default:
+                parser_state.state = sbp_parser_state_t::WAITING;
+                break;
+            }
+    }
+}
+
+
+//INVARIANT: A fully received message with correct CRC is currently in parser_state
+void
+AP_GPS_SBP2::_sbp_process_message() {
+
+    //Here, we copy messages into local structs.
+    switch (parser_state.msg_type) {
+        case SBP_HEARTBEAT_MSGTYPE:
+            last_heartbeat = *((struct sbp_heartbeat_t*)parser_state.msg_buff);
+            last_heartbeat_received_ms = AP_HAL::millis();
+            waiting_for_first_heartbeat = false;
+            break;
+
+        case SBP_GPS_TIME_MSGTYPE:
+            last_gps_time = *((struct sbp_gps_time_t*)parser_state.msg_buff);
+            break;
+
+        case SBP_VEL_NED_MSGTYPE:
+            last_vel_ned = *((struct sbp_vel_ned_t*)parser_state.msg_buff);
+            break;
+
+        case SBP_POS_LLH_MSGTYPE:
+            last_pos_llh = *((struct sbp_pos_llh_t*)parser_state.msg_buff);
+            break;
+
+        case SBP_DOPS_MSGTYPE:
+            last_dops = *((struct sbp_dops_t*)parser_state.msg_buff);
+            break;
+
+        default:
+            break;
+    }
+
+    // send all messages we receive to log, even if it's an unsupported message,
+    // so we can do annditional post-processing from Dataflash logs.
+    // The log mask will be used to adjust or suppress logging
+    logging_log_raw_sbp(parser_state.msg_type, parser_state.sender_id, parser_state.msg_len, parser_state.msg_buff);
+}
+
+bool
+AP_GPS_SBP2::_attempt_state_update()
+{
+
+    uint32_t now = AP_HAL::millis();
+
+    if (waiting_for_first_heartbeat)
+        return false;
+
+    if (now - last_heartbeat_received_ms > SBP_TIMEOUT_HEARTBEAT) {
+
+        state.status = AP_GPS::NO_GPS;
```

Comment:
Please don't do this. We will start researching for GPS instances if you do this, but in a quick audit I don't see anywhere that frees the existing driver instance. I'm pretty sure this will result in a memory leak. I'm running out to eat but will check more in a bit.